### ISSUE TITLE
AssertionError raised in the case of a function without inputs.

### ISSUE THREAD

**Issue by evelyn-king** [pipefunc/pipefunc#238]
I'm trying to make a pipeline with a loading function that takes no inputs. Something like the following:

```python
def func_a() -> float:
    return 10

def func_b(a: float) -> float:
    return a + 30

Pipeline(
    [
        PipeFunc(func_a, "a"),
        PipeFunc(func_b, "b")
    ]
)
```

When I execute this, I get the following assertion error:
```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[13], line 1
----> 1 Pipeline(
      2     [
      3         PipeFunc(func_a, "a"),
      4         PipeFunc(func_b, "b")
      5     ]
      6 )

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:129, in Pipeline.__init__(self, functions, lazy, debug, profile, cache_type, cache_kwargs, scope, default_resources)
    127     else:
    128         mapspec = None
--> 129     self.add(f, mapspec=mapspec)
    130 self._cache_type = cache_type
    131 self._cache_kwargs = cache_kwargs

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:212, in Pipeline.add(self, f, mapspec)
    209     f.debug = self.debug
    211 self._clear_internal_cache()  # reset cache
--> 212 self._validate()
    213 return f

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:942, in Pipeline._validate(self)
    940 _validate_scopes(self.functions)
    941 _check_consistent_defaults(self.functions, output_to_func=self.output_to_func)
--> 942 self._validate_mapspec()

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:954, in Pipeline._validate_mapspec(self)
    952         raise ValueError(msg)
    953 validate_consistent_axes(self.mapspecs(ordered=False))
--> 954 self._autogen_mapspec_axes()

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:997, in Pipeline._autogen_mapspec_axes(self)
    995 def _autogen_mapspec_axes(self) -> set[PipeFunc]:
    996     """Generate `MapSpec`s for functions that return arrays with ``internal_shapes``."""
--> 997     root_args = self.topological_generations.root_args
    998     mapspecs = self.mapspecs(ordered=False)
    999     non_root_inputs = _find_non_root_axes(mapspecs, root_args)

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/functools.py:1001, in cached_property.__get__(self, instance, owner)
    999 val = cache.get(self.attrname, _NOT_FOUND)
   1000 if val is _NOT_FOUND:
-> 1001     val = self.func(instance)
   1002     try:
   1003         cache[self.attrname] = val

File ~/.local/opt/micromamba/envs/pipeline/lib/python3.11/site-packages/pipefunc/_pipeline.py:981, in Pipeline.topological_generations(self)
    978 if not generations:
    979     return Generations([], [])
--> 981 assert all(isinstance(x, str | _Bound | _Resources) for x in generations[0])
    982 assert all(isinstance(x, PipeFunc) for gen in generations[1:] for x in gen)
    983 root_args = [x for x in generations[0] if isinstance(x, str)]

AssertionError: 
```

I can remove the assertion error by inserting a dummy variable in `func_a` but this feels like a buggy assertion.

**Comment by basnijholt** [comment]
Thanks for reporting!

I am aware of this problem actually but instead of raising a proper error I added this assert.

I will implement parameterless functions tomorrow.

The core issue is that when doing a topological sort on the graph, the parameterless function will be in the first "layer". I just need to write some code to re-insert that function into the layer just one above where its dependent functions are. (Never mind if this explanation doesn't make sense, just trying to put down my thoughts on the solution).

**Comment by basnijholt** [comment]
I just released [0.23.1](https://github.com/pipefunc/pipefunc/releases/tag/v0.23.1) where this is fixed :tada:
